[SPARK-26985][CORE] Fix "access only some column of the all of columns " for big endian architecture BUG

ketank-new opened the pull request:
continuation to https://github.com/apache/spark/pull/24788

## What changes were proposed in this pull request?

Changes are related to BIG ENDIAN system
This changes are done to

identify s390x platform.
use byteorder to BIG_ENDIAN for big endian systems
changes for 2 are done in access functions putFloats() and putDouble()

## How was this patch tested?

Changes have been tested to build successfully on s390x as well x86 platform to make sure build is successful.

ketank-new committed:
changes specific to s390x, this changes help in reading float and double values as exected on big endian system

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

kiszk commented:
Is this PR a replacement of #24788?

kiszk commented:
ok to test

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11706/
Test PASSed.

SparkQA commented:
**[Test build #106463 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106463/testReport)** for PR 24861 at commit [`60ea022`](https://github.com/apache/spark/commit/60ea022d6488384a8d86ea925788c4b0c56f173a).

ketank-new commented:
> Is this PR a replacement of #24788?

yes

viirya commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -304,7 +304,7 @@ public static void throwException(Throwable t) {
-    if (arch.equals("ppc64le") || arch.equals("ppc64")) {
+    if (arch.equals("ppc64le") || arch.equals("ppc64") || arch.matches("s390x")) {
       // Since java.nio.Bits.unaligned() doesn't return true on ppc (See JDK-8165231), but
       // ppc64 and ppc64le support it
```
Not big deal, but if `s390x` is listed here because of the same reason (`java.nio.Bits.unaligned() doesn't return true on`...), to update this comment seems good.

ketank-new replied:
done changed .matches to .contains

kiszk replied:
@ketank-new Is this comment to address line 309 instead of line 307?

srowen replied:
I think the question is whether the comment also applies to s390x. If not, I think it's OK to leave. But I guess the other question is, are you saying `Bits.unaligned()` also doesn't return the correct value in some cases on this platform? if it works then this isn't so necessary. If we're not sure, OK to force it here I think.

kiszk replied:
Yes, if it is not, ok to leave.   

According to my understanding, [this comment](https://github.com/apache/spark/pull/24861/files#r294620569) did not seem to answer the original question.

viirya commented:
Btw, we don't need to add `Continuation to :` in the PR title. Simply, `[SPARK-26985][CORE] ...`

SparkQA commented:
**[Test build #106463 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106463/testReport)** for PR 24861 at commit [`60ea022`](https://github.com/apache/spark/commit/60ea022d6488384a8d86ea925788c4b0c56f173a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106463/
Test PASSed.

srowen commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -304,7 +304,7 @@ public static void throwException(Throwable t) {
     boolean _unaligned;
     String arch = System.getProperty("os.arch", "");
-    if (arch.equals("ppc64le") || arch.equals("ppc64")) {
+    if (arch.equals("ppc64le") || arch.equals("ppc64") || arch.matches("s390x")) {
```
Still doesn't need `.matches`; that's not a regex

ketank-new replied:
@srowen : I had used "arch.equals("s390x") initially and by reading ur earlier comments i understood you asking me to use "matches", that's what i did in this new PR

Do you stilll suggest to change it to some thing else?

srowen replied:
I'm suggesting `.contains` for substring matching. `.matches` is for a regex.

ketank-new replied:
> I'm suggesting `.contains` for substring matching. `.matches` is for a regex.

so should make the file change in this PR itself or do  a new PR?

srowen replied:
Just change it here, push a new commit.

ketank-new replied:
done

ketank-new committed:
Update Platform.java

changed ".contains " to ".matches" as suggested by community reviewers

ketank-new reviewed (commented):
done changes plz review

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11814/
Test PASSed.

SparkQA commented:
**[Test build #106574 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106574/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).

SparkQA commented:
**[Test build #106574 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106574/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106574/
Test FAILed.

SparkQA commented:
**[Test build #4798 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4798/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).

SparkQA commented:
**[Test build #4798 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4798/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #4799 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4799/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).

kiszk commented:
@ketank-new Did you check @viirya 's [comment](https://github.com/apache/spark/pull/24861#discussion_r293355358)?

SparkQA commented:
**[Test build #4799 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4799/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).
 * This patch passes all tests.
 * This patch **does not merge cleanly**.
 * This patch adds no public classes.

jiangxb1987 commented on common/unsafe/src/main/java/org/apache/spark/unsafe/Platform.java:
```diff
@@ -304,7 +304,7 @@ public static void throwException(Throwable t) {
     boolean _unaligned;
     String arch = System.getProperty("os.arch", "");
-    if (arch.equals("ppc64le") || arch.equals("ppc64")) {
+    if (arch.equals("ppc64le") || arch.equals("ppc64") || arch.contains("s390x")) {
```
Will it create false positive case if we compare using `contains()` ?

srowen replied:
@ketank-new it's worth asking, what arch strings are you matching here? i.e. according to https://www.ibm.com/support/knowledgecenter/en/SSYKE2_8.0.0/com.ibm.java.80.doc/diag/preface/conventions.html do you also need to match "s390"?

ketank-new replied:
we have tested our changes on 64 bit 'Z ' machine and that is what we are currently supporting so we just want to check for s390x  only and not s390

srowen replied:
Is "s390x" the only string that should trigger this? then use `.equals`. Or do these machines report variations on "s390x" for which this is a substring?

ketank-new replied:
we think "s390x" is good , no variations in string reported across "Z" machines
do u want me to change "matches" to ".equals" now?

srowen replied:
Yes, if you're just looking for 's390x', change it to `.equals`. Otherwise looking good.

ketank-new commented:
> @ketank-new Did you check @viirya 's [comment](https://github.com/apache/spark/pull/24861#discussion_r293355358)?

yes and replied

ketank-new commented:
is there anythg still awaiting for the changes to get merged?

HyukjinKwon commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11943/
Test PASSed.

SparkQA commented:
**[Test build #106710 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106710/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).

SparkQA commented:
**[Test build #106710 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106710/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106710/
Test FAILed.

kiszk commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11963/
Test PASSed.

SparkQA commented:
**[Test build #106736 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106736/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).

SparkQA commented:
**[Test build #106736 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106736/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106736/
Test FAILed.

kiszk commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/11994/
Test PASSed.

SparkQA commented:
**[Test build #106772 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106772/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).

SparkQA commented:
**[Test build #106772 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106772/testReport)** for PR 24861 at commit [`42d02aa`](https://github.com/apache/spark/commit/42d02aa56b155f0e0001e957e845a2e15c1c1cc2).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106772/
Test PASSed.

ketank-new committed:
Update Platform.java

change done as per community member suggestion
changed arch.matches("s390x") to arch.equals("s390x")

ketank-new reviewed (commented):
done and pushed

SparkQA commented:
**[Test build #106821 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106821/testReport)** for PR 24861 at commit [`6345935`](https://github.com/apache/spark/commit/6345935d78e37d934c39fc157ba28a14a7d1d277).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12038/
Test PASSed.

SparkQA commented:
**[Test build #106821 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/106821/testReport)** for PR 24861 at commit [`6345935`](https://github.com/apache/spark/commit/6345935d78e37d934c39fc157ba28a14a7d1d277).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/106821/
Test FAILed.

SparkQA commented:
**[Test build #4805 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4805/testReport)** for PR 24861 at commit [`6345935`](https://github.com/apache/spark/commit/6345935d78e37d934c39fc157ba28a14a7d1d277).

SparkQA commented:
**[Test build #4805 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4805/testReport)** for PR 24861 at commit [`6345935`](https://github.com/apache/spark/commit/6345935d78e37d934c39fc157ba28a14a7d1d277).
 * This patch **fails Java style tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #4806 has started](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4806/testReport)** for PR 24861 at commit [`6345935`](https://github.com/apache/spark/commit/6345935d78e37d934c39fc157ba28a14a7d1d277).

SparkQA commented:
**[Test build #4806 has finished](https://amplab.cs.berkeley.edu/jenkins/job/NewSparkPullRequestBuilder/4806/testReport)** for PR 24861 at commit [`6345935`](https://github.com/apache/spark/commit/6345935d78e37d934c39fc157ba28a14a7d1d277).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

srowen commented:
Merged to master. It doesn't quite cherry-pick cleanly into 2.4, but it might be a trivial adjustment to make it do so. If you're up for it, feel free to open a PR with the same change to branch 2.4.

ketank-new commented:
@all : anything more remaining to test or review for the changes to get merged?

viirya commented:
This was merged.

srowen commented:
Yes, I'm just saying you can try making the same change to branch-2.4 as well. It was merged to master. There is a merge conflict though in the older branches.

gatorsmile commented:
This is a correctness bug. @ketank-new Could you submit a PR to 2.4?

ketank-new commented:
> This is a correctness bug. @ketank-new Could you submit a PR to 2.4?

PR to 2.4 branch u mean?

srowen commented:
Yes that's right. I think the check for 's390x' arch just needs to be in a different line in 2.4. That may be all there is to it.

ketank-new commented:
> Yes that's right. I think the check for 's390x' arch just needs to be in a different line in 2.4. That may be all there is to it.

u mean to say add a new elseif - blk just for s390x?

srowen commented:
No, I mean you're trying to make exactly the same change. The code is a little different in branch 2.4, so a direct patch won't apply. However, I believe there is simply a different location where you make the one-line change to add a case for 's390x' exactly like here. It might bear double-checking the code in 2.4 to check whether any other changes are needed, but they may not be.

ketank-new commented:
@srowen : I was just preparing myself to PR into branch 2.4 as asked by you
As i updated my master i notice my s390x related changes have already been reflecting in master 
so i just wanted to re check with you
1) Is there is anything remaining from my side to make sure my s390x changes get tagged to a version?
2) Do i still need to raise a PR to spark branch 2.4?

srowen commented:
The change is in the `master` branch so it will be released with Spark 3.0 right now. It is not in `branch-2.4` or others. I'm saying it's fine to open a PR for 2.4 too, as it needs a very slightly different change, and then it could be part of 2.4. Yes, you need a new PR. This one is merged and does not cherry-pick into 2.4.
